modules/nf-core/variancepartition/dream/templates/dream.R

```diff
@@ -0,0 +1,192 @@
+    p.value            = 1,                       # P-value threshold for topTable
+    lfc                = 0,                       # Log fold-change threshold for topTable
+    confint            = FALSE,                   # Whether to compute confidence intervals in topTable
+    proportion         = 0.01,                    # Proportion for eBayes
+    stdev_coef_lim     = "0.1,4",                 # Standard deviation coefficient limits for eBayes
+    trend              = FALSE,                   # Whether to use trend in eBayes
+    robust             = FALSE,                   # Whether to use robust method in eBayes
+    winsor_tail_p      = "0.05,0.1",              # Winsor tail probabilities for eBayes
+    ddf                = "adaptive",              # 'Satterthwaite', 'Kenward-Roger', or 'adaptive'
+    reml               = FALSE,
+    formula            = "$meta.formula",         # User-specified formula (e.g. "~ + (1 | sample_number)")
+    apply_voom         = FALSE                    # Whether to apply `voomWithDreamWeights`
+)
+
+# Load external arguments to opt list
+args_opt <- parse_args("$task.ext.args")
+for (ao in names(args_opt)) {
+    if (!ao %in% names(opt)) {
+        stop(paste("Invalid option:", ao))
+    }
+    opt[[ao]] <- args_opt[[ao]]
+}
+
+# If there are no blocking factors, convert string "null" to NULL
+if (!is.null(opt\$blocking_factors) && tolower(opt\$blocking_factors) == "null") {
+    opt\$blocking_factors <- NULL
+}
+
+# If there is no formula, convert string "null" to NULL
+if (!is.null(opt\$formula) && tolower(opt\$formula) == "null") {
+    opt\$formula <- NULL
+}
+
+# Load metadata
+metadata <- read_delim_flexible(opt\$sample_file, header = TRUE, stringsAsFactors = TRUE)
+rownames(metadata) <- metadata[[opt\$sample_id_col]]
+
+# Ensure contrast variable is factor, then relevel
+if (!is.null(opt\$contrast_reference) && opt\$contrast_reference != "") {
+    metadata[[opt\$contrast_variable]] <- factor(metadata[[opt\$contrast_variable]])
+    metadata[[opt\$contrast_variable]] <- relevel(metadata[[opt\$contrast_variable]], ref = opt\$contrast_reference)
+}
+
+# Exclude samples in metadata if specified
+if (!is.null(opt\$exclude_samples_col) && !is.null(opt\$exclude_samples_values)) {
+  metadata <- metadata[!(metadata[[opt\$exclude_samples_col]] %in% opt\$exclude_samples_values), ]
+}
+
+# Load count data
+countMatrix <- read_delim_flexible(opt\$count_file, header = TRUE, stringsAsFactors = FALSE)
+rownames(countMatrix) <- countMatrix\$gene_id # count_file/matrix must have a gene_id column.
+countMatrix <- countMatrix[, rownames(metadata), drop = FALSE]
+
+# Construct model formula using user-provided formula if available; if not, default to contrast variable only
+if (!is.null(opt\$formula) && opt\$formula != "") {
+    form <- as.formula(opt\$formula)
+} else {
+    form <- as.formula(paste0("~ ", opt\$contrast_variable))
+}
+print(form)
```

Comment:
I'd keep this module simpler and always require a formula. If someone wants to construct a formula based on the contrast, they can do so at the workflow level.

Reply:
Done